In [7]:

#4.5 final task 
#1If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

import pandas as pd
import numpy as np
import os
path = 'C:\\Users\\reece\\InstaCart Basket Analysis 4.4.23\\Data\\'
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col=False)
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orderswrangledv1.csv'), index_col=False)

In [10]:
#consistency checks that were covered in 4.5 coureswork

df_ords.head()
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [12]:
#consistency checks that were covered in 4.5 coureswork
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [13]:
#consistency checks that were covered in 4.5 coureswork

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  1.048575e+06  1.048575e+06  1.048575e+06  1.048575e+06   
mean   5.242870e+05  1.710968e+06  3.156404e+04  1.717418e+01   
std    3.026977e+05  9.875774e+05  1.816695e+04  1.771094e+01   
min    0.000000e+00  6.000000e+00  1.000000e+00  1.000000e+00   
25%    2.621435e+05  8.554670e+05  1.589600e+04  5.000000e+00   
50%    5.242870e+05  1.710657e+06  3.160800e+04  1.100000e+01   
75%    7.864305e+05  2.566970e+06  4.729700e+04  2.300000e+01   
max    1.048574e+06  3.421083e+06  6.310000e+04  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        1.048575e+06       1.048575e+06           985475.000000  
mean         2.778637e+00       1.345229e+01               11.102506  
std          2.047751e+00       4.219144e+00                9.186663  
min          0.000000e+00       0.000000e+00                0.000000  
25%          1.000000e+00       1.000000e+01                4.000000  
50%          3.000000e+00       1.300000e+01                7.000000  
75%          5.000000e+00       1.600000e+01               15.000000  
max          6.000000e+00       2.300000e+01               30.000000

In [20]:
##Check for mixed-type data in your df_ords dataframe.
#The function for checking whether a dataframe contains any mixed-type columns is as follows:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [19]:
#Check for mixed-type data in your df_ords dataframe.


#The function for checking whether a dataframe contains any mixed-type columns is as follows:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [21]:
#please notice that the column "product name is no longer within the df_prods because the function that removed mixed type columns ****

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.


2) What I found was interesting was the fact that all the MIN values were 1 for all columns,we should keep an eye on this. The max values seem odd too I can't imagine something costs 99999 at a grocery store. 

If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

I found a mixed-type data within the df_prods file. The column product_name. Please refer back to line [21] and you will see I coded df_prods.describe() to display that I removed the mixed type data from df_prods




Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

5) Based on report on missing values in df_ords I found 63100 missing values in "days_since_prior_order" this could be b/c the client is ordering daily therefore there is a no value in that field. 


In [22]:
#run a code to check for missing values in df_ords

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_prior_order    63100
dtype: int64

Missing values


In [32]:
#create the dataframe w/o missing values in product name 
df_ords.shape

(1048575, 8)

In [34]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [35]:
df_ords_clean.shape

(985475, 8)

In [ ]:

Run a check for duplicate values in your df_ords data

In [36]:
# Find full duplicates for df_prods_clean
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [38]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicate values were found it appears that only missing values were found the i the df_ords 

In [40]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders4.5.csv'))

In [41]:
df_prds.to_csv(os.path.join(path, 'Prepared Data', 'products4.5.csv'))